pkg/controller/rbac/provider/binding/reconciler.go

```diff
@@ -0,0 +1,218 @@
+		r.client = ca
+	}
+}
+
+// NewReconciler returns a Reconciler of ProviderRevisions.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+
+	r := &Reconciler{
+		mgr: mgr,
+
+		// TODO(negz): Is Updating appropriate here? Probably.
+		client: resource.ClientApplicator{
+			Client:     kube,
+			Applicator: resource.NewAPIUpdatingApplicator(kube),
+		},
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	return r
+}
+
+// A Reconciler reconciles ProviderRevisions.
+type Reconciler struct {
+	client resource.ClientApplicator
+	mgr    manager.Manager
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Reconcile a ProviderRevision by creating a ClusterRoleBinding that binds a
+// provider's service account to its system ClusterRole.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	pr := &v1alpha1.ProviderRevision{}
+	if err := r.client.Get(ctx, req.NamespacedName, pr); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We
+		// don't need to take any action in that case.
+		log.Debug(errGetPR, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPR)
+	}
+
+	log = log.WithValues(
+		"uid", pr.GetUID(),
+		"version", pr.GetResourceVersion(),
+		"name", pr.GetName(),
+	)
+
+	if meta.WasDeleted(pr) {
+		// There's nothing to do if our PR is being deleted. Any ClusterRoles
+		// we created will be garbage collected by Kubernetes.
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	l := &corev1.ServiceAccountList{}
+	if err := r.client.List(ctx, l); err != nil {
+		log.Debug(errListSAs, "error", err)
+		r.record.Event(pr, event.Warning(reasonBind, errors.Wrap(err, errListSAs)))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	// Filter down to the ServiceAccounts that are owned by this
+	// ProviderRevision. Each revision should control at most one, but it's easy
+	// and relatively harmless for us to handle there being many.
+	subjects := make([]rbacv1.Subject, 0)
+	for _, sa := range l.Items {
+		for _, ref := range sa.GetOwnerReferences() {
+			if ref.UID == pr.GetUID() {
+				subjects = append(subjects, rbacv1.Subject{
+					Kind:      rbacv1.ServiceAccountKind,
+					Namespace: sa.GetNamespace(),
+					Name:      sa.GetName(),
+				})
+			}
+		}
+	}
+
+	n := roles.SystemClusterRoleName(pr.GetName())
+	ref := meta.AsController(meta.TypedReferenceTo(pr, v1alpha1.ProviderRevisionGroupVersionKind))
+	rb := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            n,
+			OwnerReferences: []metav1.OwnerReference{ref},
```

Comment:
Do we want to create these for inactive provider revisions? They will not have a controller running so they will not be made use of

Reply:
One benefit of this strategy is that we don't have to handle any cleanup, we just defer to GC. However, we also will be keeping around some extraneous cluster roles. Does it cause an issue to create a cluster role with an SA subject that does not exist anymore?

Reply:
> One benefit of this strategy is that we don't have to handle any cleanup, we just defer to GC.

Right - that's a big part of why I prefer this approach. I initially went with something more like what you're suggesting, but creating roles and bindings for all revisions means:

* Crossplane and the cluster scoped roles (`crossplane-admin`, etc) has access to edit the resource of an inactive provider - e.g. to cleanup finalizers before removing the revision.
* Roles and bindings work instantaneously when updating to an existing revision.
* No need to manually GC roles and bindings for revisions that still exist.

> Does it cause an issue to create a cluster role with an SA subject that does not exist anymore?

No, you can create a ClusterRoleBinding to any subject whether they exist or not.